### PR TITLE
refs #110 table/treetable memento improvements

### PR DESCRIPTION
**Pull request by martinbluehweis**
add capability to listen to added columns, ignore columns without id

### CODE REVIEW

File: modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/table/TableViewUtil.java [hunk 1]
```diff
@@ -94,29 +97,39 @@
...
 	 */
 	public static <S> TableView<S> restoreState(TableView<S> view, Memento m, Function<TableColumn<S, ?>, String> columnKeyProvider, TableView_Properties... properties) {
-		Map<String, TableColumn<S, ?>> idToNodeMap = view.getColumns().stream().collect(Collectors.toMap(columnKeyProvider, c -> c));
+		Map<String, TableColumn<S, ?>> idToNodeMap = view.getColumns().stream().filter(c -> columnKeyProvider.apply(c) != null).collect(Collectors.toMap(columnKeyProvider, c -> c));
 		Map<TableColumn<S, ?>, String> nodeToIdMap = idToNodeMap.entrySet().stream().collect(Collectors.toMap(Entry::getValue, Entry::getKey));
 
-		for (TableColumn<S, ?> c : view.getColumns()) {
-			String key = nodeToIdMap.get(c);
+		@SuppressWarnings({ "unchecked", "null" })
+		final @NonNull List<String> savedOrder = m.get("column_order", List.class, Collections.emptyList()); //$NON-NLS-1$
```
Comment (tomsontom) [inline modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/table/TableViewUtil.java]: I think "column_order" should be a constant not?

File: modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/table/TableViewUtil.java [hunk 2]
```diff
@@ -94,29 +97,39 @@
...
 
-			if (key == null) {
-				continue;
-			}
+		view.getColumns().addListener((ListChangeListener<TableColumn<S, ?>>) change -> {
+			change.reset();
```
Comment (tomsontom) [inline modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/table/TableViewUtil.java]: Why is reset needed here?

File: modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/table/TableViewUtil.java [hunk 3]
```diff
@@ -180,24 +217,38 @@
 	public static <S> TableView<S> captureState(TableView<S> view, Memento m, Function<TableColumn<S, ?>, String> columnKeyProvider, TableView_Properties... properties) {
 		Map<TableColumn<S, ?>, String> nodeToIdMap = view.getColumns().stream().collect(Collectors.toMap(c -> c, columnKeyProvider));
 
-		for (TableColumn<S, ?> c : view.getColumns()) {
-			String key = nodeToIdMap.get(c);
-			if (key == null) {
-				continue;
+		view.getColumns().addListener((ListChangeListener<TableColumn<S, ?>>) change -> {
+			change.reset();
```
Comment (tomsontom) [inline modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/table/TableViewUtil.java]: why is reset needed here?

File: modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/tree/TreeTableViewUtil.java [hunk 4]
```diff
@@ -94,29 +97,39 @@
...
 	 */
 	public static <S> TreeTableView<S> restoreState(TreeTableView<S> view, Memento m, Function<TreeTableColumn<S, ?>, String> columnKeyProvider, TreeTableView_Properties... properties) {
-		Map<String, TreeTableColumn<S, ?>> idToNodeMap = view.getColumns().stream().collect(Collectors.toMap(columnKeyProvider, c -> c));
+		Map<String, TreeTableColumn<S, ?>> idToNodeMap = view.getColumns().stream().filter(c -> columnKeyProvider.apply(c) != null).collect(Collectors.toMap(columnKeyProvider, c -> c));
 		Map<TreeTableColumn<S, ?>, String> nodeToIdMap = idToNodeMap.entrySet().stream().collect(Collectors.toMap(Entry::getValue, Entry::getKey));
 
-		for (TreeTableColumn<S, ?> c : view.getColumns()) {
-			String key = nodeToIdMap.get(c);
+		@SuppressWarnings({ "unchecked", "null" })
+		final @NonNull List<String> savedOrder = m.get("column_order", List.class, Collections.emptyList()); //$NON-NLS-1$
```
Comment (tomsontom) [inline modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/tree/TreeTableViewUtil.java]: use constant here

File: modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/tree/TreeTableViewUtil.java [hunk 5]
```diff
@@ -94,29 +97,39 @@
...
 
-			if (key == null) {
-				continue;
-			}
+		view.getColumns().addListener((ListChangeListener<TreeTableColumn<S, ?>>) change -> {
+			change.reset();
```
Comment (tomsontom) [inline modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/tree/TreeTableViewUtil.java]: why reset?

File: modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/table/TableViewUtil.java (hunk 2 above)
Comment (martinbluehweis) [reply modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/table/TableViewUtil.java]: workaround for jdb bug 8164096 (which is fixed in u122)

File: modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/table/TableViewUtil.java (hunk 3 above)
Comment (martinbluehweis) [reply modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/table/TableViewUtil.java]: workaround for jdk bug 8164096 (which is fixed in u122)

File: modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/tree/TreeTableViewUtil.java (hunk 5 above)
Comment (martinbluehweis) [reply modules/ui/org.eclipse.fx.ui.controls/src/main/java/org/eclipse/fx/ui/controls/tree/TreeTableViewUtil.java]: workaround for jdk bug 8164096 (which is fixed in u122)

**Pull request merged**